android/src/main/java/io/ably/flutter/plugin/AblyMessageCodec.java

```diff
@@ -355,26 +358,73 @@ private ChannelOptions decodeRestChannelOptions(Map<String, Object> jsonMap) {
     try {
-      return ChannelOptions.withCipherKey((String) cipher);
-    } catch (AblyException ae) {
-      System.out.println("Exception while decoding RestChannelOptions : " + ae);
+      return createChannelOptions(cipher);
+    } catch (AblyException e) {
+      System.out.println("Exception while decoding RestChannelOptions: " + e);
       return null;
     }
   }
 
   private ChannelOptions decodeRealtimeChannelOptions(Map<String, Object> jsonMap) {
     if (jsonMap == null) return null;
     final Object cipher = jsonMap.get(PlatformConstants.TxRealtimeChannelOptions.cipher);
+
+    ChannelOptions options;
     try {
-      final ChannelOptions o = ChannelOptions.withCipherKey((String) cipher);
-      readValueFromJson(jsonMap, PlatformConstants.TxRealtimeChannelOptions.params, v -> o.cipherParams = (Map<String, String>) v);
-      // modes is not supported in ably-java
-      // Track @ https://github.com/ably/ably-flutter/issues/14
-      return o;
-    } catch (AblyException ae) {
-      System.out.println("Exception while decoding RealtimeChannelOptions: " + ae);
+      options = createChannelOptions(cipher);
+    } catch (AblyException e) {
+      System.out.println("Exception while decoding RealtimeChannelOptions: " + e);
       return null;
     }
+    options.params = (Map<String, String>) jsonMap.get(PlatformConstants.TxRealtimeChannelOptions.params);
+    final ArrayList<String> modes = (ArrayList<String>) jsonMap.get(PlatformConstants.TxRealtimeChannelOptions.modes);
+    if (modes != null && modes.size() > 0) {
+      options.modes = createChannelModesArray(modes);
+    }
+
+    return options;
+  }
+
+  private ChannelOptions createChannelOptions(@Nullable Object cipher) throws AblyException {
+    if (cipher == null) return new ChannelOptions();
+    if (cipher instanceof String) {
+      try {
+        return ChannelOptions.withCipherKey((String) cipher);
+      } catch (AblyException ae) {
+        throw AblyException.fromErrorInfo(new ErrorInfo("Exception while decoding RealtimeChannelOptions as String: " + ae, 400, 40000));
+      }
+    } else if (cipher instanceof byte[]) {
+      try {
+        return ChannelOptions.withCipherKey((byte[]) cipher);
+      } catch (AblyException ae) {
+        throw AblyException.fromErrorInfo(new ErrorInfo("Exception while decoding RealtimeChannelOptions as byte array: " + ae, 400, 40000));
+      }
+    } else {
+      throw AblyException.fromErrorInfo(new ErrorInfo("CipherKey must either be a String or a Byte Array.", 400, 40000));
+    }
+  }
+
+  private ChannelMode[] createChannelModesArray(ArrayList<String> modesString) {
+    ChannelMode[] modes = new ChannelMode[modesString.size()];
+    for (int i = 0; i < modesString.size(); i++) {
+      modes[i] = decodeChannelOptionsMode(modesString.get(i));
+    }
+    return modes;
+  }
+
+  private ChannelMode decodeChannelOptionsMode(String string) {
```

Comment:
Can you change the name of the param from string to something else please? Can be something like `modeString` I also think you could benefit from creating a static map which maps the mode strings to modes and you could just read it from the map.

Reply:
For static map, if you haven't done it before you can use `static` bloc in the class https://stackoverflow.com/a/507658

Reply:
I've renamed it in 746100d9eb496bc30c7ec43eb12e9bfe57a450ae

However, what are the benefits of creating the static map? To me, the existing switch statement is simpler and shorter.

Reply:
As we spoke on the phone, it's a map of some constant values and applies nicely to this situation. It is common to map strings to some typed values, especially when translating from an external source using a defined protocol. The end goal is to have a map which you can use around app and that is a reference for that protocol. Anyway, in this case this function is also acceptable but I'd encourage you to refactor this for later.